File: src/compiler/UCA.py
from __future__ import annotations
from dataclasses import dataclass
# MyPy isn't picking up that this adds methods to the classes like from_json
from dataclasses_json import dataclass_json
from functools import reduce
from typing import List, Literal
import pandas as pd

def get_column_names(context: Context) -> List[str]:
    return ['uca'] + context.vars + ["Hazards"]

def flatten(l):
    return [item for sublist in l for item in sublist]

def context_to_df_row(context: Context) -> List[str]:
    return flatten([[context.name], context.values, [' '.join(context.hazard_list)]])

def contexts_to_df(contexts: List[Context]):
    if len(contexts) > 0:
        data = [context_to_df_row(context) for context in contexts]
        columns = get_column_names(contexts[0])
        df = pd.DataFrame(data, columns=columns)
        return df    
    else:
        return pd.DataFrame()

# Data structures for UCAS to match the pasta dsl ast enabling the sending of json and automatic deserialization without handwritten code
@dataclass_json
@dataclass(frozen=True)
class Context:
    hazard_list: List[str]
    name: str
    values: List[Literal["True", "False", "Unknown"]]
    vars: List[str]


def rule_to_df(rule: Rule):
    data = [[rule.name, rule.system + '.' + rule.control_action, rule.type]]
    columns=['Name ', 'Control Action', 'type']
    print(columns)
    print(data)
    df = pd.DataFrame(data, columns=['Name ', 'Control Action', 'type'])
    return df    



@dataclass_json
@dataclass(frozen=True)
class Rule: 
    control_action: str
    contexts: List[Context]    
    name: str
    system: str
    type: Literal['applied-to-long', 'not-provided', 'provided', 'stopped-too-soon', 'too-early', 'too-late', 'wrong-time']

File: src/compiler/UCA.ts
import { Rule } from "../language/generated/ast"

export type myContext = {
    hazard_list: string[]
    name: string
    values: Array<"True" | "False" | "Unknown">
    vars: Array<string>
}
export type myRule = {
    control_action: string
    contexts: Array<myContext>
    name: string
    system: string
    type: 'applied-too-long' | 'not-provided' | 'provided' | 'stopped-too-soon' | 'too-early' | 'too-late' | 'wrong-time'
}


export function rule_to_myRule(rule : Rule): myRule {
    var r: myRule = {} as myRule; 
    r.contexts = [];
    for(var context of rule.contexts){
        var c:myContext = {} as myContext;
        /* console.log(`\tname: ${context.name}`) */
        c.name = context.name;

        /* console.log(`\t\tharzard_list: ${context.list.refs}`)                     */
        c.hazard_list = context.list.refs;

        /* console.log(`\t\tvars: ${context.vars}`)                     */
        c.vars = context.vars;

        /* console.log(`\t\tvalues: ${context.values}`)                     */
        c.values = context.values
                    .map(value => {
                            if(value == "True") { return "True" }
                            else if (value == "False" ) { return "False" }
                            else { return "Unknown" }
                        }   
                        );

        r.contexts.push(c)
    }
    r = {
        name: rule.name, 
        control_action: rule.action,
        type: rule.type,
        system: rule.system,
        contexts: r.contexts
    };
    return r; 
}
